### PR TITLE
Resolve Optim Params Issues

### PR DESCRIPTION
**Pull request by rayankrish**
- Includes a test of Optimizer Parameter Groups for the ONNX BERT Model (3 variations)
- Resolves the issue of not passing default hyperparameters for parameters not in a group
- Resolves the issue of sending 'lambda_coef' instead of 'lambda' to the backend
- Resolves the issue of sending lr to the backend as a hyperparameter

### CODE REVIEW

File: orttraining/orttraining/python/experimental/orttrainer.py [hunk 1]
```diff
@@ -532,12 +532,30 @@ def _create_ort_training_session(self):
...
             optimizer_int_attributes_map[initializer.name] = {}
+            not_in_param_groups = True
             for param_group in self.optim_config.params:
                 if initializer.name not in param_group['params']:
                     continue  # keep looking for a matching param_group
+                not_in_param_groups = False
                 for k, v in param_group.items():
-                    if k == 'params':
+                    if k == 'params' or k == 'lr':
                         continue  # 'params' is not a hyper parameter, skip it
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/experimental/orttrainer.py:542]: update this comment saying multiple 'lr' per weight is not supported yet too

File: orttraining/orttraining/python/experimental/orttrainer.py [hunk 2]
```diff
@@ -532,12 +532,30 @@ def _create_ort_training_session(self):
...
+                    else:
+                        raise ValueError("Optimizer attributes must be either float or int.")
+ 
+            # set default values for params not found in groups
+            if not_in_param_groups:
+                for k, v in self.optim_config.defaults.items():
+                    if k == 'lr':
+                        continue
+                    if k == 'lambda_coef':
+                        k = 'lambda'
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/experimental/orttrainer.py:558]: Move the logic converting lambda_coef -> lambda and skipping lr from the param groups to config.py

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 3]
```diff
@@ -73,6 +73,20 @@ def bert_model_description(dynamic_shape=True):
...
+    for initializer in model.graph.initializer:
+        if any(key in initializer.name for key in no_decay_keys):
+            no_decay_param_group.append(initializer.name)
+        else:
+            decay_param_group.append(initializer.name)
+    params = [{'params': no_decay_param_group, "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.0, "epsilon": 1e-6},
+              {'params': decay_param_group, "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.01, "epsilon": 1e-6}]
+    return params
+
+def optimizer_parameters_old(model):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:88]:
```suggestion
def optimizer_parameters_1group(model):
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 4]
```diff
@@ -73,6 +73,20 @@ def bert_model_description(dynamic_shape=True):
 def optimizer_parameters(model):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:73]:
```suggestion
def optimizer_parameters_2groups(model):
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 5]
```diff
@@ -73,6 +73,20 @@ def bert_model_description(dynamic_shape=True):
...
+        if any(key in initializer.name for key in no_decay_keys):
+            no_decay_param_group.append(initializer.name)
+        else:
+            decay_param_group.append(initializer.name)
+    params = [{'params': no_decay_param_group, "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.0, "epsilon": 1e-6},
+              {'params': decay_param_group, "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.01, "epsilon": 1e-6}]
+    return params
+
+def optimizer_parameters_old(model):
+    '''A method to assign different hyper parameters for different model parameter groups'''
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:89]:
```suggestion
    '''A method to assign different hyper parameters for different model parameter groups'''

```
Always add an empty line after docstring

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 6]
```diff
@@ -911,3 +925,57 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+
+    #optim_config = optim.LambConfig(params, do_bias_correction=False)
+    optim_config = optim.LambConfig(params, alpha= 0.9, beta= 0.999, lambda_coef= 0.01, epsilon= 1e-6, do_bias_correction=False)
+    opts =  orttrainer.ORTTrainerOptions({
+        'debug' : {
+            'deterministic_compute': True
+        },
+    })
+    
+    torch.manual_seed(1)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py]: set seed = 1 in the first line of code

Copy the template from previous examples in this file. there is a "#common setup" piece

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 7]
```diff
@@ -911,3 +925,57 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    set_seed(1)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, options=opts)
+
+    experimental_losses = []
+    for i in range(total_steps):
+        sample_input = generate_random_input_from_model_desc(model_desc, i)
+        experimental_losses.append(trainer.train_step(*sample_input).cpu().item())
+        
+    # LEGACY IMPLEMENTATION
+    device = torch.device("cuda", 0)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:963]: same for device.

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 8]
```diff
@@ -911,3 +925,57 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    legacy_trainer = Legacy_ORTTrainer(model, None, legacy_model_desc, "LambOptimizer",
+                       legacy_optim_map,
+                       learning_rate_description,
+                       device,
+                       _use_deterministic_compute=True)
+    legacy_losses = []
+    for i in range(total_steps):
+        sample_input = generate_random_input_from_model_desc(model_desc, i)
+        legacy_sample_input = [*sample_input, learning_rate]
+
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:977]:
```suggestion
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 9]
```diff
@@ -911,3 +925,57 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    device = torch.device("cuda", 0)
+    legacy_model_desc, learning_rate_description, learning_rate = legacy_model_params() 
+    torch.manual_seed(1)
+    set_seed(1)
+    
+    legacy_trainer = Legacy_ORTTrainer(model, None, legacy_model_desc, "LambOptimizer",
+                       legacy_optim_map,
+                       learning_rate_description,
+                       device,
+                       _use_deterministic_compute=True)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:972]:
```suggestion
    legacy_trainer = Legacy_ORTTrainer(model, None, legacy_model_desc, "LambOptimizer",
                                                                legacy_optim_map,
                                                                learning_rate_description,
                                                                device,
                                                                _use_deterministic_compute=True)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 10]
```diff
@@ -911,3 +925,57 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    legacy_losses = []
+    for i in range(total_steps):
+        sample_input = generate_random_input_from_model_desc(model_desc, i)
+        legacy_sample_input = [*sample_input, learning_rate]
+
+        legacy_losses.append(legacy_trainer.train_step(legacy_sample_input).cpu().item())
+
+    _test_helpers.assert_model_outputs(experimental_losses, legacy_losses)
+    print(legacy_losses)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:981]:
```suggestion
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 11]
```diff
@@ -911,3 +925,58 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    ([], legacy_optim_params_a),
+    # Change the hyperparameters for a subset of hardcoded parameters
+    ([{'params':['bert.embeddings.LayerNorm.bias', 'bert.embeddings.LayerNorm.weight'], "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.0, "epsilon": 1e-6}], legacy_optim_params_b),
+    # Change the hyperparameters for a generated set of paramers
+    (optimizer_parameters_old(load_bert_onnx_model()), legacy_optim_params_c)
+])
+def testToyBERTModelLegacyExperimentalCustomOptimParameters(params, legacy_optim_map):
+    total_steps = 10
+    seed = 1
+
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:940]:
```suggestion
device = 'cuda'
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 12]
```diff
@@ -911,3 +925,58 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    # EXPERIMENTAL API
+    model_desc = bert_model_description()
+    model = load_bert_onnx_model()
+
+    #optim_config = optim.LambConfig(params, do_bias_correction=False)
+    optim_config = optim.LambConfig(params, alpha= 0.9, beta= 0.999, lambda_coef= 0.01, epsilon= 1e-6, do_bias_correction=False)
+    opts =  orttrainer.ORTTrainerOptions({
+        'debug' : {
+            'deterministic_compute': True
+        },
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:950]:
```suggestion
        'device' : {
             'id' : device
        },
        'debug' : {
            'deterministic_compute': True
        },
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 13]
```diff
@@ -911,3 +925,58 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    onnxruntime.set_seed(seed)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, options=opts)
+
+    experimental_losses = []
+    for i in range(total_steps):
+        sample_input = generate_random_input_from_model_desc(model_desc, i)
+        experimental_losses.append(trainer.train_step(*sample_input).cpu().item())
+        
+    # LEGACY IMPLEMENTATION
+    device = torch.device("cuda", 0)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:963]:
```suggestion
    device = torch.device(device)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 14]
```diff
@@ -911,3 +926,60 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+                                                          learning_rate_description,
+                                                          device,
+                                                          _use_deterministic_compute=True)
+    legacy_losses = []
+    for i in range(total_steps):
+        sample_input = generate_random_input_from_model_desc(model_desc, i)
+        legacy_sample_input = [*sample_input, learning_rate]
+        legacy_losses.append(legacy_trainer.train_step(legacy_sample_input).cpu().item())
+
+    _test_helpers.assert_model_outputs(experimental_losses, legacy_losses)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:984]:
```suggestion
     # Check results
    _test_helpers.assert_model_outputs(experimental_losses, legacy_losses)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 15]
```diff
@@ -911,3 +926,60 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+@pytest.mark.parametrize("params, legacy_optim_map", [
+    # Change the hyper parameters for all parameters
+    ([], legacy_optim_params_a),
+    # Change the hyperparameters for a subset of hardcoded parameters
+    ([{'params':['bert.embeddings.LayerNorm.bias', 'bert.embeddings.LayerNorm.weight'], "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.0, "epsilon": 1e-6}], legacy_optim_params_b),
+    # Change the hyperparameters for a generated set of paramers
+    (optimizer_parameters(load_bert_onnx_model()), legacy_optim_params_c)
+])
+def testToyBERTModelLegacyExperimentalCustomOptimParameters(params, legacy_optim_map):
+    total_steps = 10
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:939]:
```suggestion
    # Common setup
    total_steps = 10
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 16]
```diff
@@ -911,3 +926,61 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    # Common setup
+    total_steps = 10
+    device = "cuda"
+    seed = 1
+
+    # EXPERIMENTAL API
+    model_desc = bert_model_description()
+    model = load_bert_onnx_model()
+
+    #optim_config = optim.LambConfig(params, do_bias_correction=False)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:948]:
```suggestion
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 17]
```diff
@@ -911,3 +926,60 @@ def testToyBERTModelGradientAccumulationLegacyExperimental(gradient_accumulation
...
+    device = torch.device(device)
+    legacy_model_desc, learning_rate_description, learning_rate = legacy_model_params(trainer.optim_config.lr) 
+    torch.manual_seed(seed)
+    onnxruntime.set_seed(seed)
+    
+    legacy_trainer = Legacy_ORTTrainer(model, None, legacy_model_desc, "LambOptimizer",
+                                                          legacy_optim_map,
+                                                          learning_rate_description,
+                                                          device,
+                                                          _use_deterministic_compute=True)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:977]:
```suggestion
    legacy_trainer = Legacy_ORTTrainer(model, None, legacy_model_desc, "LambOptimizer",
                                       legacy_optim_map,
                                       learning_rate_description,
                                       device,
                                       _use_deterministic_compute=True)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py [hunk 18]
```diff
@@ -70,9 +70,24 @@ def bert_model_description(dynamic_shape=True):
...
+    for initializer in model.graph.initializer:
+        if any(key in initializer.name for key in no_decay_keys):
+            no_decay_param_group.append(initializer.name)
+        else:
+            decay_param_group.append(initializer.name)
+    params = [{'params': no_decay_param_group, "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.0, "epsilon": 1e-6},
+              {'params': decay_param_group, "alpha": 0.9, "beta": 0.999, "lambda_coef": 0.01, "epsilon": 1e-6}]
+    return params
+
+def optimizer_parameters(model):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_bert_toy_onnx.py:88]:
```suggestion

def optimizer_parameters(model):
```
